File: front/apps/web-antd/src/api/userecho/tenant-rbac.ts
/**
 * 租户角色与成员管理 API
 */

import { requestClient } from '#/api/request';

/** 租户角色 */
export interface TenantRole {
    id: string;
    tenant_id: string;
    name: string;
    code: string;
    description: string | null;
    is_builtin: boolean;
    sort: number;
    status: string;
}

/** 租户权限 */
export interface TenantPermission {
    id: string;
    parent_id: string | null;
    name: string;
    code: string;
    type: string;
    sort: number;
}

/** 租户成员 */
export interface TenantMember {
    id: string;
    tenant_id: string;
    user_id: number;
    user_type: string;
    status: string;
    feedback_count: number;
    // 前端补充字段
    username?: string;
    nickname?: string;
    email?: string;
    roles?: TenantRole[];
}

/** 创建角色参数 */
export interface CreateRoleParams {
    name: string;
    code: string;
    description?: string;
    permission_ids?: string[];
}

/** 更新角色参数 */
export interface UpdateRoleParams {
    name?: string;
    description?: string;
    status?: string;
}

/** 创建成员参数 */
export interface CreateMemberParams {
    username: string;
    nickname: string;
    password: string;
    email?: string;
    role_ids?: string[];
}

/** 更新成员参数 */
export interface UpdateMemberParams {
    user_type?: string;
    status?: string;
}

// ==================== 角色 API ====================

/** 获取角色列表 */
export async function getTenantRoles(status?: string) {
    return requestClient.get<TenantRole[]>('/api/v1/app/tenant/roles', {
        params: { status },
    });
}

/** 获取权限列表 */
export async function getTenantPermissions() {
    return requestClient.get<TenantPermission[]>('/api/v1/app/tenant/roles/permissions');
}

/** 获取角色详情 */
export async function getTenantRole(roleId: string) {
    return requestClient.get<TenantRole>(`/api/v1/app/tenant/roles/${roleId}`);
}

/** 获取角色权限 */
export async function getTenantRolePermissions(roleId: string) {
    return requestClient.get<TenantPermission[]>(`/api/v1/app/tenant/roles/${roleId}/permissions`);
}

/** 创建角色 */
export async function createTenantRole(data: CreateRoleParams) {
    return requestClient.post<TenantRole>('/api/v1/app/tenant/roles', data);
}

/** 更新角色 */
export async function updateTenantRole(roleId: string, data: UpdateRoleParams) {
    return requestClient.put<TenantRole>(`/api/v1/app/tenant/roles/${roleId}`, data);
}

/** 更新角色权限 */
export async function updateTenantRolePermissions(roleId: string, permissionIds: string[]) {
    return requestClient.put(`/api/v1/app/tenant/roles/${roleId}/permissions`, {
        permission_ids: permissionIds,
    });
}

/** 删除角色 */
export async function deleteTenantRole(roleId: string) {
    return requestClient.delete(`/api/v1/app/tenant/roles/${roleId}`);
}

// ==================== 成员 API ====================

/** 获取成员列表 */
export async function getTenantMembers(status?: string) {
    return requestClient.get<TenantMember[]>('/api/v1/app/tenant/members', {
        params: { status },
    });
}

/** 获取我的权限 */
export async function getMyPermissions() {
    return requestClient.get<string[]>('/api/v1/app/tenant/members/my-permissions');
}

/** 获取成员详情 */
export async function getTenantMember(memberId: string) {
    return requestClient.get<TenantMember>(`/api/v1/app/tenant/members/${memberId}`);
}

/** 获取成员角色 */
export async function getTenantMemberRoles(memberId: string) {
    return requestClient.get<string[]>(`/api/v1/app/tenant/members/${memberId}/roles`);
}

/** 创建成员 */
export async function createTenantMember(data: CreateMemberParams) {
    return requestClient.post<TenantMember>('/api/v1/app/tenant/members', data);
}

/** 更新成员 */
export async function updateTenantMember(memberId: string, data: UpdateMemberParams) {
    return requestClient.put<TenantMember>(`/api/v1/app/tenant/members/${memberId}`, data);
}

/** 更新成员角色 */
export async function updateTenantMemberRoles(memberId: string, roleIds: string[]) {
    return requestClient.put(`/api/v1/app/tenant/members/${memberId}/roles`, {
        role_ids: roleIds,
    });
}

/** 删除成员 */
export async function deleteTenantMember(memberId: string) {
    return requestClient.delete(`/api/v1/app/tenant/members/${memberId}`);
}

File: server/backend/app/userecho/schema/tenant_rbac.py
from pydantic import BaseModel, Field


class TenantRoleCreate(BaseModel):
    """创建租户角色"""

    name: str = Field(..., max_length=64, description="角色名称")
    code: str = Field(..., max_length=64, description="角色代码")
    description: str | None = Field(None, max_length=500, description="角色描述")
    permission_ids: list[str] | None = Field(None, description="权限 ID 列表")


class TenantRoleUpdate(BaseModel):
    """更新租户角色"""

    name: str | None = Field(None, max_length=64, description="角色名称")
    description: str | None = Field(None, max_length=500, description="角色描述")
    status: str | None = Field(None, description="状态")


class TenantRolePermissionUpdate(BaseModel):
    """更新角色权限"""

    permission_ids: list[str] = Field(..., description="权限 ID 列表")


class TenantRoleOut(BaseModel):
    """租户角色输出"""

    id: str
    tenant_id: str
    name: str
    code: str
    description: str | None
    is_builtin: bool
    sort: int
    status: str

    class Config:
        from_attributes = True


class TenantPermissionOut(BaseModel):
    """租户权限输出"""

    id: str
    parent_id: str | None
    name: str
    code: str
    type: str
    sort: int

    class Config:
        from_attributes = True


class TenantMemberCreate(BaseModel):
    """创建租户成员"""

    email: str = Field(..., max_length=256, description="邮箱")
    nickname: str = Field(..., max_length=64, description="昵称")
    password: str = Field(..., min_length=6, max_length=32, description="密码")
    username: str | None = Field(None, max_length=64, description="用户名(可选，用于显示)")
    role_ids: list[str] | None = Field(None, description="角色 ID 列表")


class TenantMemberUpdate(BaseModel):
    """更新租户成员"""

    username: str | None = Field(None, max_length=64, description="用户名")
    nickname: str | None = Field(None, max_length=64, description="昵称")
    role_ids: list[str] | None = Field(None, description="角色 ID 列表")
    user_type: str | None = Field(None, description="用户类型")
    status: str | None = Field(None, description="状态")


class TenantMemberRoleUpdate(BaseModel):
    """更新成员角色"""

    role_ids: list[str] = Field(..., description="角色 ID 列表")


class TenantMemberOut(BaseModel):
    """租户成员输出"""

    id: str
    tenant_id: str
    user_id: int
    user_type: str
    status: str
    feedback_count: int
    username: str | None = None
    nickname: str | None = None
    email: str | None = None
    roles: list[TenantRoleOut] = []

    class Config:
        from_attributes = True
